sdk/iot/Azure.Iot.Hub.Service/tests/DevicesClientTests.cs

```diff
@@ -2,66 +2,393 @@
+            }
+            finally
+            {
+                await Cleanup(client, devicesAndTwins.Keys);
+            }
+        }
+
+        /// <summary>
+        /// Test query by getting all twins.
+        /// For the purpose of this test, we will create multiple devices (and device twins as a byproduct)
+        /// and list all twins and verify the query returns everything expected.
+        /// </summary>
+        [Test]
+        public async Task DevicesClient_Query_GetTwins()
+        {
+            string testDeviceprefix = $"bulkDevice";
+
+            IEnumerable<DeviceIdentity> devices = BuildMultipleDevices(testDeviceprefix, BULK_DEVICE_COUNT);
+
+            IoTHubServiceClient client = GetClient();
+
+            try
+            {
+                // Create all devices
+                Response<BulkRegistryOperationResponse> createResponse = await client.Devices.CreateIdentitiesAsync(devices).ConfigureAwait(false);
+
+                Assert.IsTrue(createResponse.Value.IsSuccessful, "Bulk device creation must be successful");
+
+                // We will retry the operation since it can take some time for the query to match what was recently created.
+                int matchesFound = 0;
+                DateTimeOffset startTime = DateTime.UtcNow;
+
+                while (DateTime.UtcNow - startTime < _queryMaxWaitTime)
                 {
-                    // If the Device doesn't exist, we will create it.
-                    device = (await client.Devices.CreateOrUpdateIdentityAsync(
-                        new Models.DeviceIdentity
+                    matchesFound = 0;
+                    AsyncPageable<TwinData> twins = client.Devices.GetTwinsAsync();
+
+                    // We will verify we have twins for all recently created devices.
+                    await foreach (TwinData twin in twins)
+                    {
+                        if (devices.Any(d => d.DeviceId.Equals(twin.DeviceId, StringComparison.OrdinalIgnoreCase)))
                         {
-                            DeviceId = testDeviceName
-                        }).ConfigureAwait(false)).Value;
+                            matchesFound++;
+                        }
+                    }
+
+                    if (matchesFound == BULK_DEVICE_COUNT)
+                    {
+                        break;
+                    }
+
+                    await Task.Delay(_queryRetryInterval);
                 }
 
-                // Perform another GET. We expect the device to exist.
-                Response<Models.DeviceIdentity> response = await client.Devices.GetIdentityAsync(testDeviceName).ConfigureAwait(false);
-                device = response.Value;
-                response.GetRawResponse().Status.Should().Be(200);
+                matchesFound.Should().Be(BULK_DEVICE_COUNT, "Number of matching devices must be equal to the number of recently created devices.");
```

Comment:
The test only fails at this point if the query API failed to return the full set of twins that you just created. Maybe the error message should be "Timed out waiting for all the bulk created devices to be queryable"?

Reply:
I like that ! I will append this error message as well.